### ISSUE TITLE
Cannot use query string as part of an upstream URL

### ISSUE THREAD

**Issue by chkohner** [oauth2-proxy/oauth2-proxy#455]
It's currently impossible to forward requests to an upstream host that requires an access token in the query string (or to provide any query string values at all). For example, in instances where you want to access a private upstream origin server using an access token.

## Expected Behavior

An upstream URL should be used as-is, appending whatever path (and additional query args) are provided in the original request when forwarding on to the upstream source.

Using an upstream like follows:

`OAUTH2_PROXY_UPSTREAMS=http://private.me.com/?t=123abc`

I expect the following requests to map as follows (note the merging of query args in the second example):
* `http://[proxyurl]/file.html` --> `http://private.me.com/file.html?t=123abc`
* `http://[proxyurl]/api/foo?q=Bar` --> `http://private.me.com/api/foo?t=123abc&q=Bar`

## Current Behavior

The query string is dropped entirely from upstream URLs. For example,

`OAUTH2_PROXY_UPSTREAMS=http://private.me.com/?t=123abc`

Will forward `http://[proxyurl]/file.html` to `http://private.me.com/file.html` (without the `t=123abc` part), which in the case of an access token requirement, results in a Forbidden response for any URL requested.

## Context

Without retention of the query string, it's impossible to grant the proxy access to privately hosted content via an access token. There are probably other useful scenarios, but that's the primary use case I have encountered.

- Version used: master (362cdf7713b7f3cb25647e6afd94f370d7c6cd7d)

**Comment by steakunderscore** [comment]
How would you propose a conflict in parameters is handled? For example with an upstream of `http://private.me.com/file.html?t=123abc`, would this happen:
```
http://[proxyurl]/file.html?t=987zxy --> http://private.me.com/file.html?t=123abc
```
or:
```
http://[proxyurl]/file.html?t=987zxy --> http://private.me.com/file.html?t=987zxy
```

**Comment by chkohner** [comment]
I believe the proper response is to simply concatenate them into:

```
http://[proxyurl]/file.html?t=123abc&t=987zxy
```

and leave it to the upstream HTTP implementation to make sense of it.

**Comment by JoelSpeed** [comment]
I agree with @chkohner, I don't believe there is a standard way of handling this but I have seen many different projects approach this in different ways, including some expecting multiple entries and using the values as a list

**Comment by steakunderscore** [comment]
Agreed, that makes sense :smile:

**Comment by chkohner** [comment]
Not stale.

**Comment by ezraw** [comment]
We'd love to see this addressed. +1 for concatenation.

**Comment by JoelSpeed** [comment]
I think we could add a `params` field to the `Upstream` struct within our `options` package that could then be used by the `http` upstream proxy to add params to each request. For now we would need to convert from the legacy upstreams by parsing he `--upstream=/foo?a=b` using `url.Parse` (I think we are already doing this), then taking the parsed params from that

**Comment by djdevin** [comment]
Is this fixed by #696? We aren't seeing the issue anymore.

**Comment by JoelSpeed** [comment]
This change seems unrelated to the description in this issue. I think this issue is more that users would like to be able to specify static query strings that are added to all requests to the upstream servers. The change there shouldn't affect this code at all.
